src/lights.js

```diff
@@ -4,84 +4,79 @@
 // the light's position.
 
-import {
-	AmbientLight,
-	Color,
-	DirectionalLight,
-	Group,
-	PointLight,
-	SpotLight,
-	Vector3
-} from '../vendors/three.js';
+import { Vector3 } from '../vendors/three.js';
 
 import { scaleCoordinate } from './coordinateUtils.js';
 
-export function getInitialLightPosition(coordinate, radius, extent) {
-	// initial light position in spherical polar coordinates
-	const temporaryPosition = new Vector3(
-		...(coordinate[0] ?? scaleCoordinate(coordinate[1], extent))
-	);
-
-	const result = {
-		radius: radius * temporaryPosition.length(),
-		phi: 0,
-		theta: 0
+export function getLightingStrcut() {
+	return {
+		ambientLightColor: { value: [0, 0, 0] },
+		directionalLights: {
+			value: [],
+			properties: {
+				direction: {}, // vec3
+				color: {}      // vec3
+			}
+		},
+		pointLights: {
+			value: [],
+			properties: {
+				color: {},   // vec3
+				position: {} // vec4
+			}
+		},
+		spotLights: {
+			value: [],
+			properties: {
+				color: {},     // vec3
+				coneCos: {},   // float
+				direction: {}, // vec3
+				position: {}   // vec4
+			}
+		}
 	};
-
-	if (temporaryPosition.length() !== 0) {
-		// This code is here to avoid 0/0 division.
-		result.phi = Math.atan2(temporaryPosition.y, temporaryPosition.x) % (2 * Math.PI);
-		result.theta = Math.asin(temporaryPosition.z / result.radius);
-	}
-
-	return result;
-}
-
-export function positionLights(lights, theta, phi, focus) {
-	lights.forEach((light) =>
-		light.position.set(
-			light.radius * Math.sin(theta + light.theta) * Math.cos(phi + light.phi) + focus.x,
-			light.radius * Math.sin(theta + light.theta) * Math.sin(phi + light.phi) + focus.y,
-			light.radius * Math.cos(theta + light.theta) + focus.z
-		)
-	);
 }
 
 export default {
-	ambient: ({ color }) => {
-		return new AmbientLight(new Color(...color).getHex());
-	},
-	directional: ({ color }) => {
-		return new DirectionalLight(new Color(...color).getHex());
+	ambient: ({ color }, lights) => {
+		lights.ambientLightColor.value[0] += color[0];
+		lights.ambientLightColor.value[1] += color[1];
+		lights.ambientLightColor.value[2] += color[2];
 	},
-	spot: ({ angle = 1.57079632679, color, coords, target }, extent) => {
-		const light = new SpotLight(new Color(...color).getHex());
-
-		light.position.set(
+	directional: ({ color, coords }, lights, extent) => {
+		const direction = new Vector3(
 			...(coords[0] ?? scaleCoordinate(coords[1], extent))
-		);
-
-		light.angle = angle;
-
-		light.target.position.set(
-			...(target[0] ?? scaleCoordinate(target[1], extent))
-		);
-
-		const group = new Group();
-
-		group.add(light);
-		// We need to add the target to the scene so the its matrixWorld is updated in every camera move.
-		group.add(light.target);
-
-		return group;
+		).sub(new Vector3(
+			(extent.xmax + extent.xmin) / 2,
+			(extent.ymax + extent.ymin) / 2,
+			(extent.zmax + extent.zmin) / 2
+		)).normalize();
+
+		lights.directionalLights.value.push({
```

Comment:
I'm not sure if I should return a new object here and assign it in `graphics3d.js`.

```js
return {
    ...lights,
    directionalLights: {
        value: [
            ...directionalLights.value,
            {
                color,
                direction
            }
        ]
    }
}
```

I'm not sure if this is a side-effect, but I think it is too much trouble to avoid this.
@rocky what do you think?

Reply:
@rocky I think it should be kept as is because of performance.